gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+        if state.request.is_none() {
+            Err(Error::GpioLineRequestNotConfigured)
+        } else {
+            state.buffer = Some(EdgeEventBuffer::new(1).map_err(Error::GpiodFailed)?);
+
+            state
+                .request
+                .as_ref()
+                .unwrap()
+                .reconfigure_lines(&config)
+                .map_err(Error::GpiodFailed)
+        }
+    }
+
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()> {
+        // While waiting here for the interrupt to occur, it is possible that we receive another
+        // request to disable the interrupt instead, via a call to set_irq_type().
+        //
+        // That request shouldn't wait for this wait_for_interrupt() function to finish, else we
+        // may never proceed with set_irq_type() as the parent wait_for_interrupt() function will
+        // take the lock right after dropping one and hence set_irq_type() will wait for ever.
+        //
+        // To avoid this situation, implement the state.request as an "Arc" instance instead and
+        // take a reference to it without locking the state. With this, while we are waiting for
+        // the interrupt, set_irq_type() can come and set the state.request to None and we will be
+        // able to quit the wait loop.
+        let request = {
+            let state = &self.state[gpio as usize].write().unwrap();
+
+            match &state.request {
+                Some(x) => x.clone(),
+                None => return Err(Error::GpioIrqNotEnabled),
+            }
+        };
+
+        // Wait for the interrupt for a second.
+        match request.edge_event_wait(Duration::new(1, 0)) {
+            Err(LibGpiodError::OperationTimedOut) => return Err(Error::GpioIrqOpTimedOut),
+            x => x.map_err(Error::GpiodFailed)?,
+        }
+
+        // The interrupt has already occurred, we can lock now just fine.
+        let state = &self.state[gpio as usize].write().unwrap();
+        if let Some(buffer) = &state.buffer {
+            request
+                .edge_event_read(buffer, 1)
+                .map_err(Error::GpiodFailed)?;
+
+            Ok(())
+        } else {
+            Err(Error::GpioLineRequestNotConfigured)
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct LineState {
+    dir: u8,
+    val: Option<u16>,
+    irq_type: u16,
+}
+
+#[derive(Debug)]
+pub(crate) struct GpioController<D: GpioDevice> {
+    config: VirtioGpioConfig,
+    device: D,
+    line_state: Vec<RwLock<LineState>>,
+    line_names: String,
+}
+
+impl<D: GpioDevice> GpioController<D> {
+    // Creates a new controller corresponding to `device`.
+    pub(crate) fn new(device: D) -> Result<GpioController<D>> {
+        let ngpio = device.get_num_lines()? as u16;
+
+        // Allocate enough capacity in the beginning, we can always extend it later.
+        let mut line_names = String::with_capacity((ngpio * 10).into());
```

Comment:
What format is expected for individual line names?  It would be nice to have that down in the comment.  Otherwise people reading the code will wonder, as I did, if 10 characters is enough.

Reply:
Sure.